### PR TITLE
updated to 13.0.2, added pragmas

### PR DESCRIPTION
**Pull request by brentschmaltz**
Snapped to version 13.0.2.
The majority of the changes we #nullable enable as 13.0.2 uses nullable refs.
The other changes were suppressing warnings.

The changes in the source is the result of internal changes.
To see what those are you can compare these changes to the newtonsoft repo.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs [hunk 1]
```diff
@@ -77,7 +80,12 @@ public override string Get(char[] key, int start, int length)
             hashCode -= hashCode >> 5;
 
             // make sure index is evaluated before accessing _entries, otherwise potential race condition causing IndexOutOfRangeException
-            var index = hashCode & _mask;
+#if NET20 || NET35 || NET40
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs]: >NET35 || NET40 [](http://example.com/codeflow?start=13&length=14)

do we need these?

File: src/Microsoft.IdentityModel.Tokens/opensource/json/JsonTextWriter.Async.cs [hunk 2]
```diff
@@ -131,7 +134,34 @@ internal async Task DoCloseAsync(CancellationToken cancellationToken)
...
+        {
+            if (_writeBuffer != null)
+            {
+                BufferUtils.ReturnBuffer(_arrayPool, _writeBuffer);
+                _writeBuffer = null;
+            }
+
+            if (CloseOutput && _writer != null)
+            {
+#if HAVE_ASYNC_DISPOABLE
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/JsonTextWriter.Async.cs]: >DISPOABLE [](http://example.com/codeflow?start=15&length=9)

spelling: DISPOSABLE  -> missing an S

File: src/Microsoft.IdentityModel.Tokens/opensource/json/JsonValidatingReader.cs [hunk 3]
```diff
@@ -41,6 +41,8 @@
 
 #endif
 
+#nullable disable
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/JsonValidatingReader.cs:44]: remove?

File: src/Microsoft.IdentityModel.Tokens/opensource/json/SerializationBinder.cs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 ﻿
-#if (DOTNET || PORTABLE40 || PORTABLE || NET_CORE)
+#if (DOTNET || PORTABLE40 || PORTABLE)
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/SerializationBinder.cs]: >PORTABLE40 || PORTABLE [](http://example.com/codeflow?start=15&length=22)

do we need these?

File: src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryType.cs [hunk 5]
```diff
@@ -25,6 +25,8 @@
 
 using System;
 
+#nullable disable
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryType.cs:28]: needed?

File: src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryWriter.cs [hunk 6]
```diff
@@ -29,6 +29,8 @@
 using System.Text;
 using Microsoft.IdentityModel.Json.Utilities;
 
+#nullable disable
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryWriter.cs:32]: >nullable disable [](http://example.com/codeflow?start=1&length=16)

?

**Comment by jennyf19** [comment]
> using System;

missing license info

---
Refers to: src/Microsoft.IdentityModel.Tokens/opensource/json/Linq/JsonSelectSettings.cs:1 in 6f7d57b. [](commit_id = 6f7d57b334d4e1eccde228fa7501ced8deabc315, deletion_comment = False)

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: src/Microsoft.IdentityModel.Tokens/opensource/json/JsonTextWriter.Async.cs (hunk 2 above)
Comment (jennyf19) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/JsonTextWriter.Async.cs]: https://github.com/JamesNK/Newtonsoft.Json/pull/2698

Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/JsonTextWriter.Async.cs]: I changed it, to be consistent with the repo, but the define in M.IM.Tokens.csproj was also misspelled.
So, i changed them both.

File: src/Microsoft.IdentityModel.Tokens/opensource/json/SerializationBinder.cs (hunk 4 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/SerializationBinder.cs]: We use NET_CORE in our test code, i don't think it is related to the product code, but i will add it back.
PORTABLE40 and PORTABLE are not defined in our builds, and are in many locations in the newtonsoft source.

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs (hunk 1 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs]: We don't need any of those defines

**Comment by TimHannMSFT** [comment]
> 

This and other files seem to be changed only in adding a trailing new line. Is that necessary?

---
Refers to: src/Microsoft.IdentityModel.Tokens/opensource/json/ConstructorHandling.cs:44 in 9934704. [](commit_id = 993470408518521410b888f972f474d8a43f4b68, deletion_comment = False)

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs [hunk 7]
```diff
@@ -53,12 +55,13 @@ public DefaultJsonNameTable()
...
         /// <summary>
-        /// Gets the string containing the same characters as the specified range of characters in the given array.
+        /// Gets a string containing the same characters as the specified range of characters in the given array.
         /// </summary>
         /// <param name="key">The character array containing the name to find.</param>
         /// <param name="start">The zero-based index into the array specifying the first character of the name.</param>
         /// <param name="length">The number of characters in the name.</param>
-        public override string Get(char[] key, int start, int length)
+        /// <returns>A string containing the same characters as the specified range of characters in the given array.</returns>
+        public override string? Get(char[] key, int start, int length)
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs:64]: >string? [](http://example.com/codeflow?start=24&length=7)

huh, TIL about nullable reference types... [https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-reference-types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-reference-types) #Resolved

File: src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryType.cs (hunk 5 above)
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryType.cs:28]: >disable [](http://example.com/codeflow?start=10&length=7)

Do you know why this is required here and elsewhere? This seems to be needed where nullable is enabled by default but then in the /json/ dir, there's #endable tags which make it seem as though it's disabled by default.

**Review by TimHannMSFT — APPROVED** [review]
:shipit:

**Comment by TimHannMSFT** [comment]
Reviewed all the files for the first 3rd then spot checked the rest. LGTM

File: src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryType.cs (hunk 5 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/Bson/BsonBinaryType.cs:28]: That code is in the 13.0.2 branch and not in our last snapshot.

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs (hunk 7 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs:64]: That is the code in the 13.0.2 branch.

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs (hunk 1 above)
Comment (dannybtsai) [reply src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs]: Both are in the System.Threading and apply to all platforms we support.

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs [hunk 8]
```diff
@@ -119,7 +128,7 @@ public string Add(string key)
             hashCode -= hashCode >> 5;
             for (Entry entry = _entries[hashCode & _mask]; entry != null; entry = entry.Next)
             {
-                if (entry.HashCode == hashCode && entry.Value.Equals(key))
+                if (entry.HashCode == hashCode && entry.Value.Equals(key, StringComparison.Ordinal))
```
Comment (dannybtsai) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs:127]: >StringComparison.Ordinal [](http://example.com/codeflow?start=74&length=24)

The should not be necessary, the default is StringComparison.Ordinal already.

File: src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs [hunk 9]
```diff
@@ -158,7 +167,12 @@ private void Grow()
                 }
             }
             _entries = newEntries;
-            _mask = newMask;
+
+#if NET20 || NET35 || NET40
```
Comment (dannybtsai) [inline src/Microsoft.IdentityModel.Tokens/opensource/json/DefaultJsonNameTable.cs:167]: Like read, we probably don't need this, maybe just use Volatile.Write()?

**Pull request merged**